vim: Change logic in subword motion subword_start calculation

VVill-ga opened the pull request:
Closes #23344

Previous logic would consume a subword until it left a non-alphanumeric character (consuming any punctuation). With the new logic, the only punctuation we consume is `_` for `snake_case`, and should immediately exit after leaving - consuming only trailing underscores. This new behavior resolves the inconsistencies pointed out in the issue above, and matches what I've experienced with the CamelCaseMotion vim extension.

[Demonstration of issue examples behavior resolved](https://github.com/user-attachments/assets/d877eaff-6b65-4ca7-87c6-9dba4f850a90)

[Walking through subwords](https://github.com/user-attachments/assets/b7e5694a-6bc1-4a19-8c8e-e9ddb06e8fec)


Release Notes:

- Fixed vim mode's subword motions to properly stop at punctuation and handle `snake_case`, `camelCase`, and `dot.separated` identifiers consistently, matching vim's CamelCaseMotion behavior

VVill-ga committed:
Change logic in subword motion subword_start calculation

dinocosta committed:
Merge branch 'main' into vvill-ga/fix_subword_punctuation

dinocosta committed:
test(vim): add tests for next and previous subword start

Update `crates/vim/src/motion.rs` with two tests,
`test_next_subword_start` and `test_previous_subword_start`, in order to
verify that our implementation matches the behavior of vim's
CamelCaseMotion script – https://github.com/bkad/CamelCaseMotion .

At the moment these tests are failing and are meant to serve as a
guideline for us to fix the implementations so that the tests pass.

dinocosta reviewed (commented):
Hey @VVill-ga! 👋

Thanks for taking a stab at this. I think we need a bit more work to match CamelCaseMotion’s behavior. While reviewing your changes, I tested a range of scenarios and noticed that in some cases Zed doesn’t behave the same. I’ve added tests with the expected behavior; they’re currently failing so we can adjust the implementation to make them pass. Please feel free to run through these scenarios too and let me know if I’ve misunderstood anything.

If you’re available to keep iterating, or if you need help, let me know. Otherwise, I can take a pass at fixing 🙂

Lastly, I merged `main` into this branch. Recent migrations meant you couldn’t `cargo run` here if you’d already compiled a more recent version. Thanks!

VVill-ga committed:
Update subword detection to better handle separator characters and
stopping at punctuation

VVill-ga committed:
Fix missing character in test_previous_subword_start

dinocosta committed:
test(vim): tests for next and previous subword end

Add two new tests, `test_previous_subword_end` and
`test_next_subword_end`, to ensure that those motions are also being
tested against vim's CamelCaseMotion implemenentation.

Since the updated code was not yet passing those tests, the
implementation has also been updated to ensure that those tests now
pass.

dinocosta committed:
refactor(vim): extract subword calculation to helper functions

dinocosta committed:
refactor(vim): use shared subword helpers in object.rs and add tests

- Update `in_subword` and `around_subword` in object.rs to use the
  shared `is_subword_start` and `is_subword_end` helpers from motion.rs
- Add comprehensive test for subword text objects covering `c i w`, `d i
  w`, `c a w`, and `d a w` with snake_case, camelCase, and dot.separated
  identifiers
